11-payment-encoding.md

```diff
@@ -0,0 +1,339 @@
+Data Part, with a trailing byte containing the recovery ID (0, 1, 2 or
+3).
+
+## Rationale
+
+The 5-bit `version` fits neatly into base32 encoding (0 == `q`), and
+allows for future incompatible upgrades.  `signature` covers the
+pre-base32 encoded data for simplicity, especially since the data part
+may not end on an 8-bit boundary.  It also allows public key recovery,
+so the identity of the payee node is implied.
+
+## Tagged Fields
+
+Each Tagged Field is of format:
+
+1. `type` (5 bits)
+1. `length` (10 bits, big-endian)
+1. `data` (length x 5 bits)
+
+Currently defined Tagged Fields are:
+
+* `d` (13): short description of purpose of payment,  e.g. '1 cup of coffee'
+* `h` (23): 256-bit description of purpose of payment (SHA256).  This is used to commit to an associated description which is too long to fit, such as may be contained in a web page.
+* `x` (6): `expiry` time in seconds (big-endian). Default is 3600 (1 hour) if not specified.
+* `f` (9): fallback on-chain address.  For bitcoin, this is 5 bits of
+      witness version followed by a witness program, with witness
+      version `3` (17) meaning P2PKH, `j` (18) meaning [P2SH](https://github.com/bitcoin/bips/blob/master/bip-0016.mediawiki); both are followed by a 20-byte hash value.
+* `r` (3): extra routing information.  This should be appended to the route
+      to allow routing to non-public nodes; there may be more than one of these.
+   * `pubkey` (264 bits)
+   * `channel_id` (64 bits)
+   * `fee` (64 bits, big-endian)
+   * `cltv_expiry_delta` (16 bits, big-endian)
+
+### Requirements
+
+A writer MUST NOT include more than one `d`, `h`, or `x` fields, and
+MAY include more then one `f` field.
+
+A writer MUST include either a `d` or `h` field, and MUST NOT include
+both.  If included, a writer SHOULD make `d` a complete description of
+the purpose of the payment.  If included, a writer MUST make the preimage
+of the hashed description in `h` available through some unspecified means,
+which SHOULD be a complete description of the purpose of the payment.
+
+A writer SHOULD use the minimum `x` field length possible.
+
+If a writer offers more than one `f` field, it SHOULD specify
+preferred addresses first.  For bitcoin payments, a writer MUST set an
+`f` field to a valid witness version and program, or `17` followed by
+a public key hash, or `18` followed by a script hash.
+
+A writer MUST include at least one `r` field if it does not have a
+public channel associated with its public key.  The `pubkey` is the
+node ID of the start of the channel, `channel_id` is the channel ID
+field to identify the channel, `fee` is the total fee required to use
+that channel to send `amount` to the final node, specified in 10^-11
+currency units, and `cltv_expiry_delta` is the block delta required
+by the channel.  A writer MAY include more than one `r` field to
+indicate a sequence of non-public channels to traverse.
+
+A writer MUST pad fields to a multiple of 5 bits, using zeroes.
```

Comment:
This seems to be consistently padding with trailing bits, but should probably be made explicit. Another thing that confused me initially was that the term "field" is used to refer both to the d/h/x/f/r "tagged field" as a whole, as well as individual "fields" in the "tagged field". 

> A writer MUST include at least one `r` field (...) `channel_id` is the channel ID field to identify the channel

The individual "fields" are not padded as far as I can tell in the examples, while the "tagged field" as a whole is, and the distinction is important for "tagged fields" with multiple "fields".